server/src/main/java/org/opensearch/identity/Subject.java

```diff
@@ -29,4 +30,24 @@ public interface Subject {
      * throws SubjectDisabled
      */
     void authenticate(final AuthToken token);
+
+    /**
+     * runAs allows the caller to create a session as this subject
+     *
+     * @return A session to run transport actions in the context of this subject
+     */
+    Session runAs();
+
+    /**
+     * This construct represents a session for this subject. A session is a short-lived block
+     * where transport actions are executed as this subject
+     *
+     * @opensearch.api
+     */
+    @FunctionalInterface
+    @PublicApi(since = "2.17.0")
+    interface Session extends AutoCloseable {
```

Comment:
@reta @peternied I'm still waiting for CI results, but what do you think about the changes in here? A subject now has a method called `runAs` which returns an AutoCloseable "Session". Every plugin is given a `PluginSubject` which is provided to them by default and initialized in Node.java before `createComponents` is called.

`try (Subject.Session session = pluginSubject.runAs()) { ... }` will be a drop in replacement for `try (ThreadContext.StoredContext ctx = threadContext.stashContext()) { ... }`

Reply:
@cwperks why do we need this session? I would expect something like that:

```
T runAs(Callable<T> callable) throws Exception;
```

```
pluginSubject.runAs(() -> {
  ...
});
```

The `ThreadContext.StoredContext ctx = threadContext.stashContext()` or similar could be done by each `Subject` impl out of the box, before / after running the `callable` (so the thread context manipulation are going to be completely oblivious to plugin developers).

Reply:
Building off of reta's comment, let work backwards from the plugin experience what do they have today vs what could it look like with this change.